dashboard/app/models/code_review.rb

```diff
@@ -0,0 +1,71 @@
+# == Schema Information
+#
+# Table name: code_review_requests
+#
+#  id              :bigint           not null, primary key
+#  user_id         :integer          not null
+#  script_id       :integer          not null
+#  level_id        :integer          not null
+#  project_id      :integer          not null
+#  project_version :string(255)      not null
+#  closed_at       :datetime
+#  deleted_at      :datetime
+#  created_at      :datetime         not null
+#  updated_at      :datetime         not null
+#
+# Indexes
+#
+#  index_code_review_requests_unique  (user_id,script_id,level_id,closed_at,deleted_at) UNIQUE
+class CodeReview < ApplicationRecord
+  # This model was renamed partway through the development process. This line
+  # will be removed when the table is renamed before this work is completed.
+  self.table_name = 'code_review_requests'
+
+  acts_as_paranoid
+
+  # TODO: Once all the renaming has settled, the following association should be:
+  # has_many :comments, class_name: 'CodeReviewComment', dependent:  :destroy
+  has_many :comments, class_name: 'CodeReviewNote', foreign_key: 'code_review_request_id', dependent:  :destroy
+
+  # Enforce that each student can only have one open code review per script and
+  # level. (This is also enforced at the database level with a unique index.)
+  validates_uniqueness_of :user_id, scope: [:script_id, :level_id],
+    conditions: -> {where(closed_at: nil)},
+    message: 'already has an open code review for this script and level'
+
+  # Returns an array of all code reviews (open and closed) that match the given
+  # identifying attributes. Returns an empty array if there are no matches.
+  def self.get_all(user_id:, script_id:, level_id:)
+    CodeReview.where(user_id: user_id, script_id: script_id, level_id: level_id).to_a
+  end
+
+  # Returns whether the code review is open for adding more comments.
+  def open?
+    closed_at.nil?
+  end
+
+  # Marks the code review as closed. Caller must call `save` to commit the
+  # change to the database.
+  def close
+    self.closed_at = DateTime.now
```

Comment:
Is it worth checking if the review is open here? (totally up to you, I don't have a strong opinion)

Reply:
Yeah, I went back and forth on the error handling here.  From your and Ben's responses, I think I'll try to tighten this up a bit.  I'll probably still have the API return success if the caller tries to close an already-closed review (because the end-state is what the user desired) but I'll add an error if the user tries to pass `isClosed=false`